components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/impersonation/notifiers/ImpersonationEmailNotifier.java

```diff
@@ -0,0 +1,124 @@
+    private static final String USER_NAME = "user-name";
+    private static final String LOGIN_TIME = "login-time";
+    private static final String IMPERSONATION_USER_NAME = "impersonator-user-name";
+    private static final String TEMPLATE_TYPE_IMPERSONATION = "ImpersonationEmailNotification";
+    private static final String TEMPLATE_TYPE = "TEMPLATE_TYPE";
+
+    /**
+     * Triggers a notification event when an impersonation occurs. This method checks if email notifications are
+     * enabled for the tenant domain and, if so, sends an email notification with the impersonation details.
+     *
+     * @param subjectId The ID of the user being impersonated.
+     * @param impersonatorId The ID of the impersonator.
+     * @param tenantDomain The domain of the tenant where the impersonation occurred.
+     * @param clientId The client id.
+     */
+    public void triggerNotification(String subjectId, String impersonatorId, String tenantDomain, String clientId) {
+
+        try {
+            boolean sendEmail = isSendEmail(tenantDomain);
+
+            if (sendEmail) {
+                String eventName = IdentityEventConstants.Event.TRIGGER_NOTIFICATION;
+
+                ZonedDateTime nowUtc = ZonedDateTime.now(ZoneId.of(UTC));
+
+                DateTimeFormatter formatter = DateTimeFormatter.ofPattern(DATE_FORMAT,
+                        Locale.ENGLISH);
+
+                String formattedDateTime = nowUtc.format(formatter);
+
+                AuthenticatedUser subject = OAuth2Util.getAuthenticatedUser(subjectId, tenantDomain, clientId);
+                AuthenticatedUser impersonator = OAuth2Util.getAuthenticatedUserFromSubjectIdentifier(impersonatorId,
+                        tenantDomain, clientId);
+
+                HashMap<String, Object> properties = new HashMap<>();
+                properties.put(USER_NAME, subject.getUserName());
+                properties.put(LOGIN_TIME, formattedDateTime);
+                properties.put(IMPERSONATION_USER_NAME, impersonator.getUserName());
+                properties.put(IdentityEventConstants.EventProperty.TENANT_DOMAIN, subject.getTenantDomain());
+                properties.put(IdentityEventConstants.EventProperty.USER_STORE_DOMAIN, subject.getUserStoreDomain());
+                properties.put(TEMPLATE_TYPE, TEMPLATE_TYPE_IMPERSONATION);
+
+                Event identityMgtEvent = new Event(eventName, properties);
+                OAuth2ServiceComponentHolder.getIdentityEventService()
+                        .handleEvent(identityMgtEvent);
+            }
+        } catch (Exception e) {
+            String errorMsg = "Error occurred while calling triggerNotification, detail : " + e.getMessage();
+            /*
+            We are not throwing any exception from here, because this event notification should not break the main
+            flow.
+             */
+            LOG.warn(errorMsg);
+            if (LOG.isDebugEnabled()) {
+                LOG.debug(errorMsg, e);
```

Comment:
Do we need to add both warning and debug log ? How about adding only to the warn log?